### PR TITLE
Extend websocket server to support message limits and throttling

### PR DESCRIPTION
**Pull request by iche033**
## Summary

Added 3 new features:

* setting max msg queue size
    * Msg published to websocket clients can now have a msg queue size limit to prevent the queue from infinitely growing. You can now set this by specifying:
    ```
      <queue_size_per_connection>10</queue_size_per_connection>
    ```
    * see https://github.com/osrf/subt/pull/950
    *  To test, run web app, and navigate to the websocket visualizer page. Connect to a point cloud topic, then use browser debugger to pause the page and verify that memory usage does not keep growing.

* throttling publish rate for topics
   * Added a keyword `throttle` that can be used by clients to limit the publish rate of a specific topic. This is initiated by the websocket client and it affects all clients who are subscribed to that topic.
   * see related [pull request](https://gitlab.com/ignitionrobotics/web/app/-/merge_requests/122) in web app

* limiting number of subscriptions per msg type
    * some msg types have large message size. So this configuration allows users to limit the number of subscriptions that each connection can open for a particular msg type, e.g. the following limits each client to subscribe to only 1 point cloud topic

    ```
    <subscription_limit_per_connection>
      <subscription>
        <msg_type>ignition.msgs.PointCloudPacked</msg_type>
        <limit>1</limit>
      </subscription>
    </subscription_limit_per_connection>
    ```
    * see https://github.com/osrf/subt/pull/950
    * To test, run web app, go to cloudsim's websocket visualizer, and try to connect to multiple point cloud topics. With the above configuration, only one can be visualized.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: plugins/websocket_server/WebsocketServer.hh [hunk 1]
```diff
@@ -175,6 +175,18 @@ namespace ignition
...
+
+      /// \brief Check and update subscription count for a message type. If
+      /// a client has more subscriptions to a topic of a specified type than
+      /// the subscription limit, this will block subscription. On the other
+      /// hand, for an unsubscription operation, the count is decremented.
+      /// \param[in] _topic Topic to subscribe to or unsubscribe from
+      /// \param[in] _socketId Connection socket id
+      /// \param[in] _subscribe True for subscribe operation, false for
+      /// unsubscribe operation
+      public: bool UpdateMsgTypeSubscriptionCount(const std::string &_topic,
```
Comment (ahcorde) [inline plugins/websocket_server/WebsocketServer.hh:187]: \return

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: plugins/websocket_server/WebsocketServer.hh (hunk 1 above)
Comment (iche033) [reply plugins/websocket_server/WebsocketServer.hh:188]: ebc1c5eb088827a49e2bbbd9a53f3e00df7466d2

File: plugins/websocket_server/WebsocketServer.cc [hunk 2]
```diff
@@ -924,10 +1027,18 @@ void WebsocketServer::OnWebsocketSubscribedMessage(
...
-        if (this->connections.find(socketId) != this->connections.end())
+        auto conIt = this->connections.find(socketId);
+        if (conIt != this->connections.end())
         {
-          this->QueueMessage(this->connections[socketId].get(),
-              msg.c_str(), msg.length());
+          // do additional throttling based on client connection setting
+          auto lastPubTimeCon = conIt->second->topicTimestamps[_info.Topic()];
+          std::chrono::nanoseconds timeDeltaCon = systemTime - lastPubTimeCon;
+          if (timeDeltaCon > conIt->second->topicPublishPeriods[_info.Topic()])
```
Comment (nkoenig) [inline plugins/websocket_server/WebsocketServer.cc:1036]:
```suggestion
          if (timeDeltaCon >= conIt->second->topicPublishPeriods[_info.Topic()])
```

**Review by nkoenig — APPROVED** [review]
One minor comment

**Pull request merged**